cpp/src/arrow/record_batch.cc

```diff
@@ -301,12 +318,8 @@ namespace {
 
 Status ValidateBatch(const RecordBatch& batch, bool full_validation) {
   for (int i = 0; i < batch.num_columns(); ++i) {
+    RETURN_NOT_OK(ValidateColumnLength(batch, i));
```

Comment:
nit: passing `array` created in line 304 to `ValidateColumnLength` to avoid creating shared_ptr twice.